
HLApers
=======

Getting started
---------------

### Install required software

##### 1. HLApers

    git clone https://github.com/genevol-usp/HLApers.git

##### 2. Install R v3.4+

##### 3. In R, install the following packages

-   from CRAN:

        install.packages(c("devtools", "tidyverse", "stringdist"))

-   from Bioconductor:

<!-- -->

    if (!requireNamespace("BiocManager", quietly = TRUE))
        install.packages("BiocManager")

    BiocManager::install("Biostrings")

-   from GitHub:

        devtools::install_github("genevol-usp/hlaseqlib")

##### 4. For STAR-Salmon-based pipeline, install:

-   STAR v2.5.3a+

-   Salmon v0.8.2+

-   samtools 1.3+

-   seqtk

##### 5. For kallisto-based pipeline, install:

-   kallisto

### Download data:

##### 1. IMGT database

    git clone https://github.com/ANHIG/IMGTHLA.git

##### 2. Gencode:

-   transcripts fasta (e.g., [Gencode v30 fasta](ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_30/gencode.v30.transcripts.fa.gz))

-   corresponding annotations GTF (e.g., [Gencode v30 GTF](ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_30/gencode.v30.annotation.gtf.gz))

HLApers usage
-------------

### Getting help

HLApers is composed of the following modes:

``` bash
./hlapers --help
```

    Usage: hlapers [modes]

    prepare-ref          Prepare transcript fasta files.
    index                Create index for read alignment.
    genotype             Infer HLA genotypes.

##### 1. Building a transcriptome supplemented with HLA sequences

The first step is to use `hlapers prepare-ref` to build an index composed of Gencode transcripts, where we replace the HLA transcripts with IMGT HLA allele sequences.

``` bash
./hlapers prepare-ref --help
```

    Usage: hlapers prepare-ref [options]

    -t | --transcripts   Fasta with Gencode transcript sequences.
    -a | --annotations   GTF from Gencode for the same Genome version.
    -i | --imgt          Path to IMGT directory.
    -o | --out           hladb directory.

Example:

    ./hlapers prepare-ref -t gencode.v25.transcripts.fa.gz -a gencode.v25.annotation.gtf.gz -i IMGTHLA -o hladb

##### 2. Creating an index for read alignment

``` bash
./hlapers index --help
```

    Usage: hlapers index [options]

    -t | --transcripts   Fasta with Gencode transcript sequences.
    -p | --threads       Number of threads.
    -o | --out           Output directory.

Example:

    ./hlapers index -t hladb/gencode_MHC_HLAsupp.fa -p 4 -o index

##### 3. HLA genotyping

``` bash
./hlapers genotype --help
```

    Usage: hlapers genotype [options]

    -i | --index         Index generated by 'hlapers index'.
    -t | --transcripts   Fasta with Gencode transcripts sequences used for 'hlapers index'.
    -m | --mhccoords     Genomics coordinates of the MHC region.
    -b | --bam           BAM file with Genomic alignments.
    -p | --threads       Number of threads.
    -o | --outprefix     Output prefix name.

Example:

    ./hlapers genotype -i index/STARMHC -t ./hladb/gencode_MHC_HLAsupp.fa -m hladb/mhc_coords.txt -b ERR188021.bam -p 8 -o results/ERR188021
